yash-executor/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+///
+/// `Executor` implements `Clone` but all clones share the same set of tasks.
+/// Separately created `Executor` instances do not share tasks.
+#[derive(Clone, Debug, Default)]
+pub struct Executor<'a> {
+    state: Rc<RefCell<ExecutorState<'a>>>,
+}
+
+/// Interface for spawning tasks
+///
+/// `Spawner` provides a subset of the functionality of `Executor` to allow
+/// spawning tasks without access to the full executor.
+///
+/// `Spawner` instances can be cloned and share the same executor state.
+/// `Spawner`s maintain a weak reference to the executor state, so they do not
+/// prevent the executor from being dropped. If the executor is dropped, the
+/// `Spawner` will not be able to spawn any more tasks.
+///
+/// To obtain a `Spawner` from an `Executor`, use the [`Executor::spawner`]
+/// method. The [`dead`](Self::dead) and `default` functions return a `Spawner`
+/// that can never spawn tasks.
+///
+/// ```
+/// # use yash_executor::Executor;
+/// let executor = Executor::new();
+/// let spawner = executor.spawner();
+/// let final_receiver = unsafe {
+///     executor.spawn(async move {
+///         let receiver_1 = spawner.spawn(async { 1 }).unwrap();
+///         let receiver_2 = spawner.spawn(async { 2 }).unwrap();
+///         receiver_2.await + receiver_1.await
+///     })
+/// };
+/// executor.run_until_stalled();
+/// assert_eq!(final_receiver.try_receive(), Ok(3));
+/// ```
+#[derive(Clone, Debug, Default)]
+pub struct Spawner<'a> {
+    state: Weak<RefCell<ExecutorState<'a>>>,
+}
+
+/// Internal state of the executor
+#[derive(Default)]
+struct ExecutorState<'a> {
+    /// Queue of woken tasks to be executed
+    ///
+    /// Tasks are added to the queue when they are woken up by another task or
+    /// when they are spawned. The executor removes a task from the queue and
+    /// polls it once. If the poll method returns `Poll::Pending`, the task
+    /// needs to be added back to the queue by some waker when it is ready to
+    /// be polled again.
+    wake_queue: VecDeque<Rc<Task<'a>>>,
+    // We don't need to store tasks that are waiting to be woken up because they
+    // are retained by wakers. This also prevents leaking tasks that are never
+    // woken up.
+}
+
+impl Debug for ExecutorState<'_> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("ExecutorState")
+            .field(
+                "wake_queue",
+                &format_args!("(len = {})", self.wake_queue.len()),
+            )
+            .finish()
+    }
+}
+
+/// State of a task to be executed
+struct Task<'a> {
+    /// Shared state of the executor for running this task
+    executor: Weak<RefCell<ExecutorState<'a>>>,
+
+    /// The task to be executed
+    ///
+    /// This value becomes `None` when the task is completed to prevent polling
+    /// it again.
+    future: RefCell<Option<Pin<Box<dyn Future<Output = ()> + 'a>>>>,
```

Comment:
Use of `RefCell` seems inevitable to me since `Task` is contained in `Rc` so that wakers containing a reference to the task can co-exist while the task is being polled which requires a mutable reference. Any ideas?